feat: add rolling upgrade support for nifi 2.x

xeniape opened the pull request:
# Description

Implementation of https://github.com/stackabletech/nifi-operator/issues/736
This PR configures the rolling upgrade for clusters upgrading from version 2.x to 2.y. It also expands the upgrade integration test.

```
--- PASS: kuttl (2670.97s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/logging_nifi-1.27.0_zookeeper-latest-3.9.3_openshift-false (184.85s)
        --- PASS: kuttl/harness/ldap_nifi-2.2.0_zookeeper-latest-3.9.3_ldap-use-tls-false_openshift-false (185.65s)
        --- PASS: kuttl/harness/smoke_nifi-1.28.1_zookeeper-3.9.3_openshift-false_listener-class-external-unstable (252.70s)
        --- PASS: kuttl/harness/cluster_operation_nifi-latest-2.2.0_zookeeper-latest-3.9.3_openshift-false (91.17s)
        --- PASS: kuttl/harness/resources_nifi-2.2.0_zookeeper-latest-3.9.3_openshift-false (84.79s)
        --- PASS: kuttl/harness/resources_nifi-1.27.0_zookeeper-latest-3.9.3_openshift-false (199.67s)
        --- PASS: kuttl/harness/resources_nifi-1.28.1_zookeeper-latest-3.9.3_openshift-false (230.88s)
        --- PASS: kuttl/harness/smoke_nifi-1.28.1_zookeeper-3.9.3_openshift-false_listener-class-cluster-internal (273.97s)
        --- PASS: kuttl/harness/upgrade_nifi_old-1.27.0_nifi_new-2.2.0_zookeeper-latest-3.9.3_openshift-false (375.19s)
        --- PASS: kuttl/harness/smoke_nifi-1.28.1_zookeeper-3.9.2_openshift-false_listener-class-external-unstable (250.60s)
        --- PASS: kuttl/harness/smoke_nifi-1.28.1_zookeeper-3.9.2_openshift-false_listener-class-cluster-internal (266.33s)
        --- PASS: kuttl/harness/smoke_nifi-1.27.0_zookeeper-3.9.3_openshift-false_listener-class-external-unstable (264.48s)
        --- PASS: kuttl/harness/smoke_nifi-1.27.0_zookeeper-3.9.3_openshift-false_listener-class-cluster-internal (244.04s)
        --- PASS: kuttl/harness/smoke_nifi-1.27.0_zookeeper-3.9.2_openshift-false_listener-class-external-unstable (229.98s)
        --- PASS: kuttl/harness/smoke_nifi-1.27.0_zookeeper-3.9.2_openshift-false_listener-class-cluster-internal (256.39s)
        --- PASS: kuttl/harness/orphaned_resources_nifi-2.2.0_zookeeper-latest-3.9.3_openshift-false (74.50s)
        --- PASS: kuttl/harness/upgrade_nifi_old-2.0.0_nifi_new-2.2.0_zookeeper-latest-3.9.3_openshift-false (341.71s)
        --- PASS: kuttl/harness/orphaned_resources_nifi-1.28.1_zookeeper-latest-3.9.3_openshift-false (222.26s)
        --- PASS: kuttl/harness/ldap_nifi-1.28.1_zookeeper-latest-3.9.3_ldap-use-tls-false_openshift-false (192.22s)
        --- PASS: kuttl/harness/ldap_nifi-1.28.1_zookeeper-latest-3.9.3_ldap-use-tls-true_openshift-false (193.35s)
        --- PASS: kuttl/harness/ldap_nifi-1.27.0_zookeeper-latest-3.9.3_ldap-use-tls-true_openshift-false (181.43s)
        --- PASS: kuttl/harness/smoke_nifi-2.2.0_zookeeper-3.9.3_openshift-false_listener-class-cluster-internal (393.69s)
        --- PASS: kuttl/harness/smoke_nifi-2.2.0_zookeeper-3.9.3_openshift-false_listener-class-external-unstable (308.61s)
        --- PASS: kuttl/harness/oidc_nifi-1.28.1_zookeeper-latest-3.9.3_oidc-use-tls-false_openshift-false (103.77s)
        --- PASS: kuttl/harness/ldap_nifi-1.27.0_zookeeper-latest-3.9.3_ldap-use-tls-false_openshift-false (185.14s)
        --- PASS: kuttl/harness/oidc_nifi-1.27.0_zookeeper-latest-3.9.3_oidc-use-tls-false_openshift-false (102.90s)
        --- PASS: kuttl/harness/oidc_nifi-2.2.0_zookeeper-latest-3.9.3_oidc-use-tls-true_openshift-false (86.37s)
        --- PASS: kuttl/harness/oidc_nifi-2.2.0_zookeeper-latest-3.9.3_oidc-use-tls-false_openshift-false (87.73s)
        --- PASS: kuttl/harness/oidc_nifi-1.27.0_zookeeper-latest-3.9.3_oidc-use-tls-true_openshift-false (106.12s)
        --- PASS: kuttl/harness/oidc_nifi-1.28.1_zookeeper-latest-3.9.3_oidc-use-tls-true_openshift-false (112.98s)
        --- PASS: kuttl/harness/logging_nifi-2.2.0_zookeeper-latest-3.9.3_openshift-false (96.08s)
        --- PASS: kuttl/harness/orphaned_resources_nifi-1.27.0_zookeeper-latest-3.9.3_openshift-false (235.80s)
        --- PASS: kuttl/harness/logging_nifi-1.28.1_zookeeper-latest-3.9.3_openshift-false (177.56s)
        --- PASS: kuttl/harness/smoke_nifi-2.2.0_zookeeper-3.9.2_openshift-false_listener-class-external-unstable (319.63s)
        --- PASS: kuttl/harness/ldap_nifi-2.2.0_zookeeper-latest-3.9.3_ldap-use-tls-true_openshift-false (173.58s)
        --- PASS: kuttl/harness/smoke_nifi-2.2.0_zookeeper-3.9.2_openshift-false_listener-class-cluster-internal (419.51s)
PASS
```

The `upgrade_nifi_old-2.0.0_nifi_new-2.2.0_zookeeper-latest-3.9.3_openshift-false` integration test was run with a custom built nifi image `2.0.0-stackable0.0.0-dev` since we currently don't have a release with two separate nifi 2.x versions. If this changes, the tests will need some adjustments, for example in `test-definition.yaml` for the `nifi_old` dimension. This is currently left out to not break current test runs.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

maltesander commented on rust/operator-binary/src/controller.rs:
```diff
@@ -1416,7 +1460,11 @@ async fn build_node_rolegroup_statefulset(
             template: pod_template,
             update_strategy: Some(StatefulSetUpdateStrategy {
-                type_: Some("OnDelete".to_string()),
+                type_: if version_change_state.is_none() {
```
I would use a boolean as parameter here like `can_rolling_update` and do the none check when calling this.

maltesander reviewed (changes requested):
Tests work fine, flows keep working (and running) from 1.27.0 to 2.0.0 to 2.2.0 and displays accordingly 
![Screenshot from 2025-04-09 16-51-19](https://github.com/user-attachments/assets/7a79b0f0-4ccb-4d73-abd9-9ca1aa1244c3)

There is one problem when i tried with the custom 2.0.0 image to the standard 2.2.0. When using a non custom image afterwards and no serverside apply the propery is still set. This then showed the nifi version in the UI to be 2.0.0 (but using 2.2.0 image).
This should be fixed in `02-install-nifi.yaml.j2` like:
```
spec:
  image:
{% if test_scenario['values']['nifi_old'].find(",") > 0 %}
    custom: "{{ test_scenario['values']['nifi_old'].split(',')[1] }}"
    productVersion: "{{ test_scenario['values']['nifi_old'].split(',')[0] }}"
{% else %}
    productVersion: "{{ test_scenario['values']['nifi_old'] }}"
    custom: null 
{% endif %}
```
and in `05-upgrade-nifi.yaml.j2` like 
```
{% if test_scenario['values']['nifi_new'].find(",") > 0 %}
    custom: "{{ test_scenario['values']['nifi_new'].split(',')[1] }}"
    productVersion: "{{ test_scenario['values']['nifi_new'].split(',')[0] }}"
{% else %}
    productVersion: "{{ test_scenario['values']['nifi_new'] }}"
    custom: null # <-----thats required
{% endif %}
```
```

xeniape commented:
> Tests work fine, flows keep working (and running) from 1.27.0 to 2.0.0 to 2.2.0 and displays accordingly ![Screenshot from 2025-04-09 16-51-19](https://private-user-images.githubusercontent.com/19609456/431925561-7a79b0f0-4ccb-4d73-abd9-9ca1aa1244c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQyNjU3MzEsIm5iZiI6MTc0NDI2NTQzMSwicGF0aCI6Ii8xOTYwOTQ1Ni80MzE5MjU1NjEtN2E3OWIwZjAtNGNjYi00ZDczLWFiZDktOWNhMWFhMTI0NGMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEwVDA2MTAzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0OTk0MzE4ODY2YTUyNjllNzYzNGEyNDgzOTY1YmQzODVjMDI1YmYxZDhhYWQxMDQyNzE0MTczNTM2NjliZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S_QajRv_TRuQMbmtTd5GGN2dLGHDyxaJmrhahmdUw8g)
> 
> There is one problem when i tried with the custom 2.0.0 image to the standard 2.2.0. When using a non custom image afterwards and no serverside apply the propery is still set. This then showed the nifi version in the UI to be 2.0.0 (but using 2.2.0 image). This should be fixed in `02-install-nifi.yaml.j2` like:
> 
> ```
> spec:
>   image:
> {% if test_scenario['values']['nifi_old'].find(",") > 0 %}
>     custom: "{{ test_scenario['values']['nifi_old'].split(',')[1] }}"
>     productVersion: "{{ test_scenario['values']['nifi_old'].split(',')[0] }}"
> {% else %}
>     productVersion: "{{ test_scenario['values']['nifi_old'] }}"
>     custom: null 
> {% endif %}
> ```
> 
> and in `05-upgrade-nifi.yaml.j2` like
> 
> ```
> {% if test_scenario['values']['nifi_new'].find(",") > 0 %}
>     custom: "{{ test_scenario['values']['nifi_new'].split(',')[1] }}"
>     productVersion: "{{ test_scenario['values']['nifi_new'].split(',')[0] }}"
> {% else %}
>     productVersion: "{{ test_scenario['values']['nifi_new'] }}"
>     custom: null # <-----thats required
> {% endif %}
> ```

Good find, thanks! Added the fix in all the integration tests https://github.com/stackabletech/nifi-operator/pull/771/commits/8b20f2af56bcb82493b493a243b5d2abb898b396

Also reran an affected test:
```
--- PASS: kuttl (408.70s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/upgrade_nifi_old-2.0.0,docker.stackable.tech_sandbox_nifi_2.0.0-stackable0.0.0-dev_nifi_new-2.2.0_zookeeper-latest-3.9.3_openshift-false (408.69s)
PASS
```

xeniape commented on docs/modules/nifi/pages/usage_guide/updating.adoc:
```diff
@@ -17,10 +17,16 @@ spec:
+====
+NiFi clusters cannot be upgraded or downgraded in a rolling fashion due to a limitation in NiFi prior to version 2.
+
+When upgrading between NiFi 1 versions or from NiFi 1 to NiFi 2, any change to the NiFi version in the CRD triggers a full cluster restart with brief downtime.
```
That's already mentioned some lines further down https://docs.stackable.tech/home/stable/nifi/usage_guide/updating/#_nifi_2_0_0 or did you mean something else?

xeniape commented on rust/operator-binary/src/controller.rs:
```diff
@@ -358,11 +358,14 @@ impl ReconcilerError for Error {
-    BeginChange,
-    Stopped,
-    NoChange,
+enum VersionChangeState {
```
I adjusted some of the variants and the enum name itself, thank you for the input! https://github.com/stackabletech/nifi-operator/pull/771/commits/827ac39e2ebbf83d07ac5b464fcaa777eab8d406
Not sure what the difference of the suggested `UpdateRequested` and `ReadyForUpdate` is , we don't distinguish between those in the code.

xeniape replied:
Ah, I think you meant to replace `ReadyForChange` with `UpdateRequested` and `NoVersionChange` with `ReadyForUpdate`? The first one would actually be better yes, changed it here again https://github.com/stackabletech/nifi-operator/pull/771/commits/450367cbedba093fda62f7733efd5864d6b1ccf2
The second one I would leave as is, since the scenarios using that state are if the specified version in nifi cluster and the deployed one are the same (so no version change detected) and if the cluster isn't started yet for the first time, so no deployed version present yet. But feel free to correct me :)

maltesander replied:
Yeah sorry that was bad quoting. I meant: 
```
    ReadyForChange -> UpdateRequested,
    InProgress -> UpdateInProgress,
    ClusterStopped -> ClusterStopped/ClusterScaleDownFinished,
    NoVersionChange -> ReadyForUpdate,
```
Im fine with leaving the second one if that is not always an update :)

xeniape commented on rust/operator-binary/src/controller.rs:
```diff
@@ -691,6 +647,91 @@ pub async fn reconcile_nifi(
     Ok(Action::await_change())
 }
 
+async fn version_change_state(
```
https://github.com/stackabletech/nifi-operator/pull/771/commits/1d15a8376822686092d5c5007cae101c99cc64c7

maltesander reviewed (approved):
LGTM!
